mllib/src/main/scala/org/apache/spark/ml/regression/AFTRegression.scala

```diff
@@ -0,0 +1,308 @@
+      rawWeights
+    }
+
+    if (handlePersistence) instances.unpersist()
+
+    val realWeights = Vectors.dense(weights.slice(2, weights.length))
+    val intercept = weights(1)
+    val scale = math.exp(weights(0))
+    val model = new AFTRegressionModel(uid, realWeights, intercept, scale)
+    copyValues(model.setParent(this))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema, fitting = true)
+  }
+
+  override def copy(extra: ParamMap): AFTRegression = defaultCopy(extra)
+}
+
+/**
+ * :: Experimental ::
+ * Model produced by [[AFTRegression]].
+ */
+@Experimental
+class AFTRegressionModel private[ml] (
+    override val uid: String,
+    val weights: Vector,
+    val intercept: Double,
+    val scale: Double)
+  extends Model[AFTRegressionModel] with AFTRegressionParams {
+
+  /** @group setParam */
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  def setQuantileCol(value: String): this.type = set(quantileCol, value)
+  setDefault(quantileCol -> "quantile")
+
+  /** @group setParam */
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  def predict(features: Vector, quantile: Vector): Vector = {
+    // scale parameter of the Weibull distribution
+    val lambda = math.exp(weights.toBreeze.dot(features.toBreeze) + intercept)
+    // shape parameter of the Weibull distribution
+    val k = 1 / scale
+    val array = quantile.toArray.map { q => lambda * math.exp(math.log(-math.log(1-q)) / k) }
+    Vectors.dense(array)
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    transformSchema(dataset.schema)
+    val predictUDF = udf { (features: Vector, quantile: Vector) => predict(features, quantile) }
+    dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol)), col($(quantileCol))))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema, fitting = false)
+  }
+
+  override def copy(extra: ParamMap): AFTRegressionModel = {
+    copyValues(new AFTRegressionModel(uid, weights, intercept, scale), extra).setParent(parent)
+  }
+}
+
+private class AFTAggregator(weights: BDV[Double], fitIntercept: Boolean)
+  extends Serializable {
+
+  private val beta = weights.slice(1, weights.length)
+  private val sigma = math.exp(weights(0))
+
+  private var totalCnt: Long = 0L
+  private var lossSum = 0.0
+  private var gradientBetaSum = BDV.zeros[Double](beta.length)
+  private var gradientLogSigmaSum = 0.0
+
+  def count: Long = totalCnt
+  def loss: Double = if (totalCnt == 0) 1.0 else lossSum / totalCnt
+  // Here we optimize loss function over beta and log(sigma)
+  def gradient: BDV[Double] = BDV.vertcat(BDV(Array(gradientLogSigmaSum / totalCnt.toDouble)),
+    gradientBetaSum/totalCnt.toDouble)
+
+  def add(data: (Vector, Double, Double)): this.type = {
+
+    val xi = if (fitIntercept) {
```

Comment:
revise depending on how you decide to set up the `initalWeights`. See comment above.